templates/e2e-dspy-workflow/README.ipynb

```diff
@@ -0,0 +1,1074 @@
+    "\n",
+    "We'll fine-tune our LLM by choosing a set of configurations. We have created recipes for different LLMs in the [`training configs`](configs/training/lora/llama-3-8b.yaml) folder which can be used as is or modified for experiments. These configurations provide flexibility over a broad range of parameters such as model, data paths, compute to use for training, number of training epochs, how often to save checkpoints, padding, loss, etc. We also include several [DeepSpeed](https://github.com/microsoft/DeepSpeed) [configurations](configs/deepspeed/zero_3_offload_optim+param.json) to choose from for further optimizations around data/model parallelism, mixed precision, checkpointing, etc.\n",
+    "\n",
+    "We also have recipes for [LoRA](https://arxiv.org/abs/2106.09685) (where we train a set of small low ranked matrices instead of the original attention and feed forward layers) or full parameter fine-tuning. We recommend starting with LoRA as it's less resource intensive and quicker to train."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 23,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Copying file:///home/ray/.dspy_cache/finetune/anyscale_6e332a9ff41cc87d.jsonl to gs://storage-bucket-cld-tffbxe9ia5phqr1unxhz4f7e1e/org_4snvy99zwbmh4gbtk64jfqggmj/cld_tffbxe9ia5phqr1unxhz4f7e1e/artifact_storage/anyscale_6e332a9ff41cc87d.jsonl\n",
+      "  \n",
+      ".\n",
+      "(anyscale +5m1.4s) Uploading local dir '.' to cloud storage.\n",
+      "(anyscale +5m4.0s) Job 'dspy-llmforge-fine-tuning-job' submitted, ID: 'prodjob_vmqx9yv5alunb2s68prunjbg8z'.\n",
+      "(anyscale +5m4.0s) View the job in the UI: https://console.anyscale.com/jobs/prodjob_vmqx9yv5alunb2s68prunjbg8z\n",
+      "(anyscale +5m4.2s) Waiting for job 'prodjob_vmqx9yv5alunb2s68prunjbg8z' to reach target state SUCCEEDED, currently in state: STARTING\n",
+      "(anyscale +10m24.2s) Job 'prodjob_vmqx9yv5alunb2s68prunjbg8z' transitioned from STARTING to RUNNING\n",
+      "(anyscale +20m30.4s) Job 'prodjob_vmqx9yv5alunb2s68prunjbg8z' transitioned from RUNNING to SUCCEEDED\n",
+      "(anyscale +20m30.4s) Job 'prodjob_vmqx9yv5alunb2s68prunjbg8z' reached target state, exiting\n"
+     ]
+    }
+   ],
+   "source": [
+    "from dspy.clients.lm import TrainingMethod\n",
+    "from src import finetuning_kwargs\n",
+    "\n",
+    "train_data = dataset_formatted\n",
+    "method = TrainingMethod.SFT\n",
+    "\n",
+    "finetuneable_lm = dspy.LM(model=\"meta-llama/Llama-3.2-1B-Instruct\")\n",
+    "\n",
+    "try:\n",
+    "    finetuning_job = finetuneable_lm.finetune(method, train_data, eval_data=None, provider=\"anyscale\", train_kwargs=finetuning_kwargs)\n",
```

Comment:
As discussed offline: I'm a little concerned about the setup here.  Looks like the `anyscale` client is using the models SDK to get the `storage_uri` field and then get model checkpoints. Currently, with `llmforge>=0.5.6` this field is gonna actually contain the path to the forwarded checkpoint according to eval loss. This is not ideal, but I also want to make sure that internal DSPy code doesn't break with a simple upgrade. 

Can we just have a manual step for retrieving the checkpoint path? It's in fact also logged to the exp. tracking tool (wandb) so we can also have a util here that can retrieve it from the run.

Reply:
Like is it the standard in DSPy fine-tuning to have this `result` object that contains information for all the checkpoints?

Reply:
My expectation was that the `.finetune` client would just interface with DSPy datasets and then submit a fine-tuning job. But maybe I'm missing context on the common API

Reply:
common API is wip. The way it currently works:
job.result() has the last (or best) checkpoint, but any time we fine-tune in DSPy we want to go over the entire list of checkpoints to find the best, so calling .result() in order to block and wait and then getting job.model_names is the way to get all the possible checkpoints